# Activity 4.01 – optimizing memory usage by converting into the appropriate dtypes

## Import the pandas package.

In [1]:
import pandas as pd

## Load the CSV file as a DataFrame.

In [2]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Pandas-Workshop/master/Chapter04/Data/car.csv'

data_frame = pd.read_csv(file_url)

## Display the first 10 rows of the DataFrame.

In [3]:
data_frame.head(10)

buying  maint  doors  persons lug_boot safety  class
0  vhigh  vhigh      2      2.0    small    low  unacc
1  vhigh  vhigh      2      2.0    small    med  unacc
2  vhigh  vhigh      2      NaN    small   high  unacc
3  vhigh  vhigh      2      2.0      med    low  unacc
4  vhigh  vhigh      2      2.0      med    med  unacc
5    NaN  vhigh      2      2.0      med   high    NaN
6  vhigh  vhigh      2      2.0      big    low  unacc
7  vhigh  vhigh      2      2.0      big    NaN  unacc
8  vhigh  vhigh      2      2.0      big   high  unacc
9  vhigh    NaN      2      4.0    small    low  unacc

## Display the data types of each column in the DataFrame using the info() method.

In [4]:
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1728 entries, 0 to 1727
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   buying    1727 non-null   object 
 1   maint     1727 non-null   object 
 2   doors     1728 non-null   int64  
 3   persons   1151 non-null   float64
 4   lug_boot  1728 non-null   object 
 5   safety    1727 non-null   object 
 6   class     1727 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 94.6+ KB


## Replace the missing values appropriately.

In [5]:
data_frame.fillna(value={'buying': 'Unknown',\
                         'maint': 'Unknown',\
                         'persons': round(data_frame.persons.mean()),\
                         'lug boot': 'Unknown',\
                         'safety': 'Unknown',\
                         'class': 'Unknown',}, inplace=True)

data_frame.head(10)

buying    maint  doors  persons lug_boot   safety    class
0    vhigh    vhigh      2      2.0    small      low    unacc
1    vhigh    vhigh      2      2.0    small      med    unacc
2    vhigh    vhigh      2      3.0    small     high    unacc
3    vhigh    vhigh      2      2.0      med      low    unacc
4    vhigh    vhigh      2      2.0      med      med    unacc
5  Unknown    vhigh      2      2.0      med     high  Unknown
6    vhigh    vhigh      2      2.0      big      low    unacc
7    vhigh    vhigh      2      2.0      big  Unknown    unacc
8    vhigh    vhigh      2      2.0      big     high    unacc
9    vhigh  Unknown      2      4.0    small      low    unacc

## Count the number of distinct unique values for buying, maint, doors, persons, lug_boot, safety, and class.

In [6]:
data_frame['buying'].nunique()

5

In [7]:
data_frame['maint'].nunique()

5

In [8]:
data_frame['doors'].nunique()

4

In [9]:
data_frame['persons'].nunique()

3

In [10]:
data_frame['lug_boot'].nunique()

3

In [11]:
data_frame['safety'].nunique()

4

In [12]:
data_frame['class'].nunique()

5

## Convert the object columns into category columns and the float64 columns into int64 columns, where appropriate, and display them:

 0   buying    1727 non-null   object >>> category

 1   maint     1727 non-null   object >>> category

 2   doors     1728 non-null   int64

 3   persons   1151 non-null   float64>>> int

 4   lug_boot  1728 non-null   object >>> category

 5   safety    1727 non-null   object >>> category

 6   class     1727 non-null   object >>> category

In [13]:
data_frame = data_frame.astype({'buying': 'category',\
                                'maint': 'category',\
                                'persons': 'int',\
                                'lug_boot': 'category',\
                                'safety': 'category',\
                                'class': 'category'})

## Finally, display the data types of each column in the DataFrame. You should see the following output:

In [14]:
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1728 entries, 0 to 1727
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   buying    1728 non-null   category
 1   maint     1728 non-null   category
 2   doors     1728 non-null   int64   
 3   persons   1728 non-null   int32   
 4   lug_boot  1728 non-null   category
 5   safety    1728 non-null   category
 6   class     1728 non-null   category
dtypes: category(5), int32(1), int64(1)
memory usage: 29.8 KB
